### PR TITLE
SN-API-36 Creating a base class serializer

### PR DESCRIPTION
**Pull request by Aretemal**

### CODE REVIEW

File: src/utils/JsonSerializer/Serializer.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Serializer {
+  serializer(data, id, type, links) {
```
Comment (Konstantsin) [inline src/utils/JsonSerializer/Serializer.js:2]: Remember methods are verbs unless it is initializer.

File: src/utils/JsonSerializer/UserSerializer.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import Serializer from './Serializer.js';
+
+class UserSerializer extends Serializer {
+  getOne(data, links) {
+    return this.serializer(data, data.id, 'ObjectUser', links);
```
Comment (Konstantsin) [inline src/utils/JsonSerializer/UserSerializer.js:5]: - Use for type jus plural form of resource name. Same for other methods.
- Do you really need many methods in serilizer?

File: src/serializers/Serializers.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      },
+    };
+  }
+
+  type() {
+    throw new Error('Method is not defined');
+  }
+
+  id() {
+    throw new Error('Method is not defined');
```
Comment (ksavon) [inline src/serializers/Serializers.js:25]: Probably u can have implementation here for all the resources?

File: src/controllers/UserController.js [hunk 4]
```diff
@@ -1,21 +1,24 @@
...
+    const serializer = new UserSerializers({
+      attributes: user, id: req.user.id, type: 'User', link: fullUrlCreator(req),
+    });
+    req.serializer = serializer;
+    next();
   }
 
-  async getAllUsers(req, res) {
+  async getAllUsers(req, res, next) {
     const data = await UserService.getAllUsers(req.user.id, req.query);
```
Comment (ksavon) [inline src/controllers/UserController.js:16]: Please name after concrete resourrce. `data` => `users`

File: src/controllers/UserController.js [hunk 5]
```diff
@@ -1,21 +1,24 @@
...
-    res.json(UserJsonCreator.getAllUsers(data, fullUrlCreator(req)));
-  }
-
-  async getStatus(req, res) {
-    const status = await UserService.getStatus(req.params.id);
-    res.json(UserJsonCreator.getStatus(status, fullUrlCreator(req)));
+    const serializer = new UserSerializers({
+      attributes: data, id: req.user.id, type: 'Users', link: fullUrlCreator(req),
```
Comment (ksavon) [inline src/controllers/UserController.js:18]: - iterate over collection of users and use serializer wih each of them.
- why not pass complete user resource into serializer?

File: src/controllers/UserController.js [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
-import UserJsonCreator from '../utils/JsonPresenters/UserJsonCreator.js';
+import UserSerializers from '../serializers/UserSerializers.js';
 
 class UserController {
-  async getOne(req, res) {
+  async getOne(req, res, next) {
     const user = await UserService.getOne(req.params.id);
-    res.json(UserJsonCreator.getOne(user, fullUrlCreator(req)));
+    const serializer = new UserSerializers({
+      attributes: user, id: req.user.id, type: 'User', link: fullUrlCreator(req),
```
Comment (ksavon) [inline src/controllers/UserController.js:9]: About type: 
https://jsonapi.org/format/#document-resource-object-identification

Check the note at the end of the section 7.2.1.

File: src/controllers/UserController.js [hunk 7]
```diff
@@ -1,21 +1,25 @@
...
-  }
-
-  async getStatus(req, res) {
-    const status = await UserService.getStatus(req.params.id);
-    res.json(UserJsonCreator.getStatus(status, fullUrlCreator(req)));
+  async getAllUsers(req, res, next) {
+    const users = await UserService.getAllUsers(req.user.id, req.query);
+    const serializer = new UsersSerializer({
+      attributes: users, link: fullUrlCreator(req),
+    }, UserSerializer);
```
Comment (Konstantsin) [inline src/controllers/UserController.js:20]: Try this way:
`new UsersSerializer(users, req)`
and for the single user:
 `new UserSerializer(user, req)`

**Pull request merged**